### PR TITLE
Replace use of os.path with pathlib

### PR DESCRIPTION
**Pull request by Icosahunter**

### CODE REVIEW

**Comment by Icosahunter** [comment]
Tested on Linux (Solus) for basic functionality

File: tagstudio/src/cli/ts_cli.py [hunk 1]
```diff
@@ -356,45 +357,44 @@ def paste_field_from_buffer(self, entry_id) -> None:
     def init_external_preview(self) -> None:
         """Initialized the external preview image file."""
         if self.lib and self.lib.library_dir:
-            external_preview_path: str = os.path.normpath(
-                f"{self.lib.library_dir}/{TS_FOLDER_NAME}/external_preview.jpg"
+            external_preview_path: Path = (
+                self.lib.library_dir / TS_FOLDER_NAME / "external_preview.jpg"
             )
-            if not os.path.isfile(external_preview_path):
+            if not external_preview_path.isfile():
```
Comment (yedpodtrzitko) [inline tagstudio/src/cli/ts_cli.py:363]:
```suggestion
            if not external_preview_path.is_file():
```

File: tagstudio/src/core/library.py [hunk 2]
```diff
@@ -350,7 +332,7 @@ def __init__(self) -> None:
         # Maps the filenames of entries in the Library to their entry's index in the self.entries list.
         #   Used for O(1) lookup of a file based on the current index (page number - 1) of the image being looked at.
         #   That filename can then be used to provide quick lookup to image metadata entries in the Library.
-        # 	NOTE: On Windows, these strings are always lowercase.
+
         self.filename_to_entry_id_map: dict[str, int] = {}
```
Comment (yedpodtrzitko) [inline tagstudio/src/core/library.py:336]:
```suggestion
        self.filename_to_entry_id_map: dict[Path, int] = {}
```

File: tagstudio/src/cli/ts_cli.py [hunk 3]
```diff
@@ -356,45 +357,44 @@ def paste_field_from_buffer(self, entry_id) -> None:
...
     def set_external_preview_default(self) -> None:
         """Sets the external preview to its default image."""
         if self.lib and self.lib.library_dir:
-            external_preview_path: str = os.path.normpath(
-                f"{self.lib.library_dir}/{TS_FOLDER_NAME}/external_preview.jpg"
+            external_preview_path: str = (
+                self.lib.library_dir / TS_FOLDER_NAME / "external_preview.jpg"
             )
-            if os.path.isfile(external_preview_path):
+            if external_preview_path.isfile():
```
Comment (yedpodtrzitko) [inline tagstudio/src/cli/ts_cli.py:374]:
```suggestion
            if external_preview_path.is_file():
```

File: tagstudio/src/cli/ts_cli.py [hunk 4]
```diff
@@ -356,45 +357,44 @@ def paste_field_from_buffer(self, entry_id) -> None:
...
                 temp.save(external_preview_path)
-
             open_file(external_preview_path)
 
     def set_external_preview_default(self) -> None:
         """Sets the external preview to its default image."""
         if self.lib and self.lib.library_dir:
-            external_preview_path: str = os.path.normpath(
-                f"{self.lib.library_dir}/{TS_FOLDER_NAME}/external_preview.jpg"
+            external_preview_path: str = (
```
Comment (yedpodtrzitko) [inline tagstudio/src/cli/ts_cli.py:371]:
```suggestion
            external_preview_path: Path = (
```

File: tagstudio/src/core/library.py [hunk 5]
```diff
@@ -1072,7 +1012,7 @@ def merge_dupe_entries(self):
                 )
             else:
                 sys.stdout.write(
-                    f"\r[LIBRARY] [{i}/{len(self.entries)}] Consolidating Duplicate: {(e.path + os.pathsep + e.filename)[0:]}..."
+                    f"\r[LIBRARY] [{i}/{len(self.entries)}] Consolidating Duplicate: {(e.path / e.filename)[0:]}..."
```
Comment (yedpodtrzitko) [inline tagstudio/src/core/library.py:1015]: > TypeError: 'PosixPath' object is not subscriptable

it needs to be converted to string before slicing it... or maybe the slice isnt even needed since it stars at the first character anyway?

File: tagstudio/src/core/library.py [hunk 6]
```diff
@@ -1114,24 +1053,16 @@ def refresh_dupe_files(self, results_filepath):
...
-                    else:
-                        if (
-                            file_1 in self.filename_to_entry_id_map.keys()
-                            and file_2 in self.filename_to_entry_id_map.keys()
-                        ):
-                            self.dupe_files.append(
-                                (files[match[0]], files[match[1]], match[2])
-                            )
+                    file_1 = str(files[match[0]].relative_to(self.library_dir))
+                    file_2 = str(files[match[1]].relative_to(self.library_dir))
```
Comment (yedpodtrzitko) [inline tagstudio/src/core/library.py:1057]: you're converting the items of `files` into strings, but using `relative_to()` here assumes it should be instance of `Path` 🤔

File: tagstudio/src/core/library.py [hunk 7]
```diff
@@ -1616,8 +1556,8 @@ def search_tags(
 
         # NOTE: I'd expect a blank query to return all with the other implementation, but
         # it misses stuff like Archive (id 0) so here's this as a catch-all.
-        query = query.strip()
-        if not query:
+
+        if query == "":
```
Comment (yedpodtrzitko) [inline tagstudio/src/core/library.py:1560]: why this change?

**Comment by yedpodtrzitko** [comment]
good job, this was really needed 👍

File: tagstudio/src/cli/ts_cli.py (hunk 1 above)
Comment (Icosahunter) [reply tagstudio/src/cli/ts_cli.py]: I think I've fixed all these.

**Comment by Icosahunter** [comment]
@yedpodtrzitko Thanks for the comments, I think I've resolved most of them. I'll have to test with the changes and ruff-format it again.

File: tagstudio/src/cli/ts_cli.py [hunk 8]
```diff
@@ -1056,9 +1053,7 @@ def create_collage(self) -> str:
                     if i < len(self.lib.entries) and run:
                         # entry: Entry = self.lib.get_entry_from_index(i)
                         entry = self.lib.entries[i]
-                        filepath = os.path.normpath(
-                            f"{self.lib.library_dir}/{entry.path}/{entry.filename}"
-                        )
+                        filepath = self.lib.library_dir / entry.path / entry.filename
                         file_type = os.path.splitext(filepath)[1].lower()[1:]
```
Comment (yedpodtrzitko) [inline tagstudio/src/cli/ts_cli.py:1057]:
```suggestion
                        file_type = filepath.suffix.lstrip('.')
```

File: tagstudio/src/core/library.py [hunk 9]
```diff
@@ -1102,7 +1041,7 @@ def refresh_dupe_files(self, results_filepath):
                 matches: list[tuple[int, int, int]] = []
                 for element in group:
                     if element.tag == "file":
-                        file = element.attrib.get("path")
+                        file = Path(element.attrib.get("path")).resolve()
                         files.append(file)
```
Comment (yedpodtrzitko) [inline tagstudio/src/core/library.py:1045]: can you change the annotation of `files` to `list[Path]` as well?

File: tagstudio/src/core/library.py [hunk 10]
```diff
@@ -1114,24 +1053,16 @@ def refresh_dupe_files(self, results_filepath):
...
-                                (files[match[0]], files[match[1]], match[2])
-                            )
+                    file_1 = str(files[match[0]].relative_to(self.library_dir))
+                    file_2 = str(files[match[1]].relative_to(self.library_dir))
+
+                    if (
+                        file_1.resolve in self.filename_to_entry_id_map.keys()
+                        and file_2 in self.filename_to_entry_id_map.keys()
+                    ):
+                        self.dupe_files.append(
```
Comment (yedpodtrzitko) [inline tagstudio/src/core/library.py:1063]: `self.dupe_files` should have changed its type annotation as well

**Comment by Icosahunter** [comment]
Fixed some issues that cropped up with the modularized GUI files.

**Comment by Icosahunter** [comment]
I think all the basics are working now, tested creating/opening library, adding tags/fields, and removing unlinked files (on Solus Linux)

**Comment by yedpodtrzitko** [comment]
running mypy, there's an error on following line, so that still needs to be adjusted:

https://github.com/TagStudioDev/TagStudio/blob/66aecf2030bea8dff84611d21b597c9b56e51c75/tagstudio/src/qt/modals/folders_to_tags.py#L70

> src/qt/modals/folders_to_tags.py:70: error: "Path" has no attribute "split"  [attr-defined]

File: tagstudio/src/core/library.py [hunk 11]
```diff
@@ -1232,14 +1163,14 @@ def _match_missing_file(self, file: str) -> list[str]:
         matches = []
 
         # for file in self.missing_files:
-        head, tail = os.path.split(file)
-        for root, dirs, files in os.walk(self.library_dir, topdown=True):
+        path = Path(file)
+        for root, dirs, files in self.library_dir.walk():
```
Comment (yedpodtrzitko) [inline tagstudio/src/core/library.py:1167]: > src/core/library.py:1167: error: "Path" has no attribute "walk"  [attr-defined]

**Review by yedpodtrzitko — CHANGES_REQUESTED** [review]

**Comment by yedpodtrzitko** [comment]
@CyanVoxel it would be nice to merge the mypy PRs sooner or later, errors like this https://github.com/TagStudioDev/TagStudio/pull/156#issuecomment-2113988948 will be caught automatically.

**Comment by CyanVoxel** [comment]
> @CyanVoxel it would be nice to merge the mypy PRs sooner or later, errors like this https://github.com/TagStudioDev/TagStudio/pull/156#issuecomment-2113988948 will be caught automatically.

I'll get that going soon here

**Comment by CyanVoxel** [comment]
Just a heads up, I'll be pulling #142 (a rewrite of #88) before this, which will cause some additional conflicts - I apologize in advance.

**Comment by yedpodtrzitko** [comment]
@CyanVoxel can you approve the CI jobs on this PR to see if everything is passing without error now?

File: tagstudio/src/qt/widgets/thumb_renderer.py [hunk 12]
```diff
@@ -275,12 +273,13 @@ def render(
                 final = ThumbRenderer.thumb_broken_512.resize(
                     (adj_size, adj_size), resample=resampling_method
                 )
-
             qim = ImageQt.ImageQt(final)
+            final.save(f'/home/icosahunter/Documents/test/{filepath.stem}.png')
```
Comment (yedpodtrzitko) [inline tagstudio/src/qt/widgets/thumb_renderer.py:277]: 🤔

Comment (Icosahunter) [reply tagstudio/src/qt/widgets/thumb_renderer.py:277]: whoops, I'll need to fix that 🤦‍♂️

**Comment by Icosahunter** [comment]
Tested basic functionality on my Linux distro (Solus) and seems to be working. At one point I did get a segfault... but it didn't happen in later testing.

File: tagstudio/src/core/library.py [hunk 13]
```diff
@@ -1091,24 +1028,23 @@ def refresh_dupe_files(self, results_filepath):
         A duplicate file is defined as an identical or near-identical file as determined
         by a DupeGuru results file.
         """
-        full_results_path = (
-            os.path.normpath(f"{self.library_dir}/{results_filepath}")
-            if self.library_dir not in results_filepath
-            else os.path.normpath(f"{results_filepath}")
-        )
-        if os.path.exists(full_results_path):
+        full_results_path = Path(results_path)
```
Comment (yedpodtrzitko) [inline tagstudio/src/core/library.py:1031]:
```suggestion
        full_results_path = Path(results_filepath)
```

File: tagstudio/src/qt/widgets/preview_panel.py [hunk 14]
```diff
@@ -499,31 +497,32 @@ def update_widgets(self):
...
+                            image = Image.open(str(filepath))
+                        elif filepath.suffix in RAW_IMAGE_TYPES:
                             with rawpy.imread(filepath) as raw:
                                 rgb = raw.postprocess()
                                 image = Image.new(
                                     "L", (rgb.shape[1], rgb.shape[0]), color="black"
                                 )
-                        elif extension in VIDEO_TYPES:
+                        elif filepath.suffix in VIDEO_TYPES:
                             video = cv2.VideoCapture(filepath)
```
Comment (CyanVoxel) [inline tagstudio/src/qt/widgets/preview_panel.py:511]:
```suggestion
                            video = cv2.VideoCapture(str(filepath))
```
cv2.VideoCapture expects a string here, and throws an error when given a Path object such as:
```
[PreviewPanel][ERROR] Couldn't Render thumbnail for F:\Example 4\Folder A\yoink\video.mp4 (because of OpenCV(4.9.0) :-1: error: (-5:Bad argument) in function 'VideoCapture'
> Overload resolution failed:
>  - Can't convert object to 'str' for 'filename'
>  - VideoCapture() missing required argument 'apiPreference' (pos 2)
>  - Argument 'index' is required to be an integer
>  - VideoCapture() missing required argument 'apiPreference' (pos 2)
)
```
Can be reproduced by checking the console log after clicking on a video file in the library. I see that the other instances outside of `ts_cli.py` are also using this cast.

File: tagstudio/src/qt/widgets/preview_panel.py [hunk 15]
```diff
@@ -499,31 +497,32 @@ def update_widgets(self):
...
                             image = Image.fromarray(frame)
 
                         # Stats for specific file types are displayed here.
-                        if extension in (IMAGE_TYPES + VIDEO_TYPES + RAW_IMAGE_TYPES):
+                        if filepath.suffix in (
+                            IMAGE_TYPES + VIDEO_TYPES + RAW_IMAGE_TYPES
+                        ):
                             self.dimensions_label.setText(
-                                f"{extension.upper()}  •  {format_size(os.stat(filepath).st_size)}\n{image.width} x {image.height} px"
+                                f"{filepath.suffix.upper()}  •  {format_size(os.stat(filepath).st_size)}\n{image.width} x {image.height} px"
```
Comment (CyanVoxel) [inline tagstudio/src/qt/widgets/preview_panel.py:522]:
```suggestion
                                f"{filepath.suffix.upper()[1:]}  •  {format_size(os.stat(filepath).st_size)}\n{image.width} x {image.height} px"
```
Removes leading "." from filetype display

File: tagstudio/src/qt/widgets/preview_panel.py [hunk 16]
```diff
@@ -499,31 +497,32 @@ def update_widgets(self):
...
                             )
                         else:
-                            self.dimensions_label.setText(f"{extension.upper()}")
+                            self.dimensions_label.setText(f"{filepath.suffix.upper()}")
```
Comment (CyanVoxel) [inline tagstudio/src/qt/widgets/preview_panel.py:525]:
```suggestion
                            self.dimensions_label.setText(f"{filepath.suffix.upper()[1:]}")
```
Removes leading "." from filetype display

File: tagstudio/src/qt/widgets/preview_panel.py [hunk 17]
```diff
@@ -535,7 +534,7 @@ def update_widgets(self):
                         DecompressionBombError,
                     ) as e:
                         self.dimensions_label.setText(
-                            f"{extension.upper()}  •  {format_size(os.stat(filepath).st_size)}"
+                            f"{filepath.suffix.upper()}  •  {format_size(os.stat(filepath).st_size)}"
```
Comment (CyanVoxel) [inline tagstudio/src/qt/widgets/preview_panel.py:537]:
```suggestion
                            f"{filepath.suffix.upper()[1:]}  •  {format_size(os.stat(filepath).st_size)}"
```
Removes leading "." from filetype display (wish I could comment on all these different lines at once)

File: tagstudio/src/qt/widgets/thumb_renderer.py [hunk 18]
```diff
@@ -289,8 +286,8 @@ def render(
                     math.ceil(adj_size / pixel_ratio),
                     math.ceil(final.size[1] / pixel_ratio),
                 ),
-                extension,
+                _filepath.suffix,
```
Comment (CyanVoxel) [inline tagstudio/src/qt/widgets/thumb_renderer.py:289]:
```suggestion
                _filepath.suffix[1:],
```
Removes leading "." from filetype display on thumbnails

File: tagstudio/src/qt/widgets/thumb_renderer.py [hunk 19]
```diff
@@ -289,8 +286,8 @@ def render(
...
             )
 
         else:
-            self.updated.emit(timestamp, QPixmap(), QSize(*base_size), extension)
+            self.updated.emit(timestamp, QPixmap(), QSize(*base_size), _filepath.suffix)
```
Comment (CyanVoxel) [inline tagstudio/src/qt/widgets/thumb_renderer.py:293]:
```suggestion
            self.updated.emit(timestamp, QPixmap(), QSize(*base_size), _filepath.suffix[1:])
```
Removes leading "." from filetype display on thumbnails

File: tagstudio/src/core/library.py (hunk 13 above)
Comment (Icosahunter) [reply tagstudio/src/core/library.py]: There were a couple other issues with that method it looks like, I fixed it up. Thanks!

**Comment by Icosahunter** [comment]
Got a segmentation fault when trying to delete unlinked entries, but I couldn't reproduce the issue.
```
[ThumbRenderer][ERROR]: Couldn't render thumbnail for /home/icosahunter/Documents/test.txt ([Errno 2] No such file or directory: '/home/icosahunter/Documents/test.txt')
Traceback (most recent call last):
  File "/home/icosahunter/Repos/TagStudio/tagstudio/src/qt/widgets/preview_panel.py", line 530, in update_widgets
    raise UnidentifiedImageError
PIL.UnidentifiedImageError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/icosahunter/Repos/TagStudio/tagstudio/src/qt/ts_qt.py", line 1209, in <lambda>
    lambda checked=False, entry=entry: self.select_item(
                                       ^^^^^^^^^^^^^^^^^
  File "/home/icosahunter/Repos/TagStudio/tagstudio/src/qt/ts_qt.py", line 1137, in select_item
    self.preview_panel.update_widgets()
  File "/home/icosahunter/Repos/TagStudio/tagstudio/src/qt/widgets/preview_panel.py", line 539, in update_widgets
    f"{filepath.suffix.upper()[1:]}  •  {format_size(os.stat(filepath).st_size)}"
                                                     ^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/home/icosahunter/Documents/test.txt'
Start RMF: <PySide6.QtCore.QThread(0x55681a5ac5b0) at 0x7fc9d3f2d600>
TagStudio.sh: line 7:  9544 Segmentation fault      (core dumped) python tagstudio/tag_studio.py
```
